src/utils/performance.js

```diff
@@ -0,0 +1,145 @@
+import Vue from 'vue'
+import { Platform } from 'quasar'
+import router from '@/base/router'
+import datastore from '@/base/datastore'
+import axios from '@/base/api/axios'
+import debounce from 'lodash.debounce'
+import { underscorizeKeys } from '@/utils/utils'
+
+const SAVE_INTERVAL_MS = 5000 // batch saves to the backend
+
+const performance = window.performance
+const fetch = window.fetch
+
+// Make sure we have all the required performance methods
+// and fetch to avoid needing the polyfill (we don't need all browser support)
+// and that vue's performance mode is not enabled (don't want to fight it)
+const ENABLED = fetch &&
+  performance &&
+  performance.clearMeasures &&
+  performance.clearMarks &&
+  performance.measure &&
+  performance.mark &&
+  !Vue.config.performance
+
+// For each load we measure up to the point where the first v-measure is measured
+// the set done to true so we don't record beyond that, until the next page load
+let done = false
+
+// The first load is the main page load from the server, after that we record the page
+// loads within the browser
+let firstLoad = true
+
+// When using performance.measure() we don't use a start mark initially, which causes it to
+// measure from the start of that whole page load, subsequent javascript page loads are only
+// measured from when we create the mark in router beforeEach()
+let startMark
+
+// What we will eventually save for this measurement run
+// Keeping it up here so its possible to add things to it as we go along...
+// ... well basically the firstLoad flag
+let currentStat = {}
+
+// Stats waiting to be saved
+const pendingStats = []
+
+function initialize () {
+  startMark = 'karrot-start'
+  done = false
+  currentStat = {}
+  performance.clearMarks()
+  performance.clearMeasures()
+  performance.clearResourceTimings()
+  performance.mark(startMark)
+}
+
+function measure (name, qualifier) {
+  if (!ENABLED || done) return
+  const label = ['karrot', name, qualifier].join(' ').trim()
+  if (startMark) {
+    performance.measure(label, startMark)
+  }
+  else {
+    performance.measure(label)
+  }
+}
+
+function save () {
+  const stats = [...pendingStats]
+  pendingStats.length = 0
+  const data = { stats }
+  // Using fetch() API instead of axios, just for this keepalive thing ...
+  fetch('/api/stats/', {
+    method: 'POST',
+    // keepalive is to help the request still work when called in the "unload" event
+    // See https://fetch.spec.whatwg.org/#request-keepalive-flag
+    keepalive: true,
+    headers: {
+      'Content-Type': 'application/json',
+      [axios.defaults.xsrfHeaderName]: readCookie(axios.defaults.xsrfCookieName),
+    },
+    body: JSON.stringify(underscorizeKeys(data)),
+  })
+}
+
+const debouncedSave = debounce(save, SAVE_INTERVAL_MS, { maxWait: SAVE_INTERVAL_MS * 2 })
```

Comment:
Or rather use `debounceAndFlushBeforeUnload` and remove the `unload` listener at the end of the file?

Reply:
Ah yeah, I went in a few cycles with it, and back to the starting point again :)